wp-content/themes/pub/wporg-learn-2020/functions.php

```diff
@@ -290,6 +290,42 @@ function wporg_archive_modify_query( WP_Query $query ) {
 }
 add_action( 'pre_get_posts', 'wporg_archive_modify_query' );
 
+/**
+ * Modify the workshop post type archive query to prioritize workshops in the user's locale.
+ *
+ * @param array    $clauses
+ * @param WP_Query $query
+ *
+ * @return array
+ */
+function wporg_archive_query_prioritize_locale( $clauses, $query ) {
+	if ( ! $query->is_post_type_archive( 'wporg_workshop' ) || is_admin() ) {
+		return $clauses;
+	}
+
+	global $wpdb;
+
+	$locale      = get_locale() ?: 'en_US';
+	$locale_root = preg_replace( '#^([a-z]{2,3}_?)[a-zA-Z_-]*#', '$1', $locale, -1, $count );
+
+	if ( $count ) {
+		$clauses['fields']  .= ", MAX( IF( {$wpdb->postmeta}.meta_key = 'video_language' AND {$wpdb->postmeta}.meta_value LIKE '{$locale_root}%', 1, 0 ) ) AS has_language";
+		$clauses['fields']  .= ", MAX( IF( {$wpdb->postmeta}.meta_key = 'video_caption_language' AND {$wpdb->postmeta}.meta_value LIKE '{$locale_root}%', 1, 0 ) ) AS has_caption";
+		$clauses['orderby'] = 'has_language DESC, has_caption DESC, ' . $clauses['orderby'];
+
+		if ( false === strpos( $clauses['join'], "INNER JOIN {$wpdb->postmeta}" ) ) {
+			$clauses['join'] .= " INNER JOIN {$wpdb->postmeta} ON ( {$wpdb->posts}.ID = {$wpdb->postmeta}.post_id )";
+		}
+
+		if ( false === strpos( $clauses['groupby'], "{$wpdb->posts}.ID" ) ) {
+			$clauses['groupby'] = "{$wpdb->posts}.ID";
+		}
+	}
```

Comment:
This would take me a relatively large amount of effort to understand if I ever needed to modify it. I'm wondering if it'd be simpler to do a custom `SELECT` query instead of trying to modify `WP_Query`?

Or maybe do a second query to grab the `video_language` / `video_caption_language` posts, and then use PHP to combine/dedupe/sort them w/ the `WP_Query` result?

If not, then some comments might help others grok it faster. In the docblock, maybe an example of the query before and after modification. And then maybe an inline comment to explain the use of `MAX()`

Reply:
I wondered if `ORDER BY FIELD` might be a clearer option, but I don't think there's a way to do that with a wildcard `LIKE`. So that would probably require storing both the specific locale and the locale root in multiple postmeta fields. And I'm not sure if `WP_Query` even supports `ORDER BY FIELD`.

Reply:
Ah, it looks like `ORDER BY CASE` might do it: https://stackoverflow.com/questions/3609166/mysql-order-by-like

But again I'm not sure that's even possible with `WP_Query`

Reply:
I added some docs in 518848d

Reply:
👍🏻